### PR TITLE
[Feature] Support DSA-CP for Hybrid scenario

### PR DESCRIPTION
**Pull request by zzhx1**
Signed-off-by: zzhx1 <zzh_201018@outlook.com>

Based on the Sharded-CP feature PR:#4702; RFC:https://github.com/vllm-project/vllm/issues/30055

### Support FULL_DECODE_ONLY Mode under PD-Mixed Scenario:
Extends DSA-CP to handle the FULL_DECODE_ONLY execution mode when running in a prefill-decode mixed (PD-mixed) serving environment, improving throughput and resource utilization for decode-intensive workloads.
**In pure prefill nodes:**
- Both q_proj and o_proj are sharded across world ranks, using **broadcast** for weights distribution.

**In PD-mixed nodes (supporting both prefill and decode):**

- q_proj is fully replicated (not sharded) to avoid communication overhead during decoding.
- o_proj Using the original TP `RowParallelLinear` method to store weights

**During prefill execution:**
- o_proj forwards through all_gather to collect weights, reconstructing the complete o_proj weights on each card.

**During decode (graph replay phase):**
- Additional all_to_all (before o_proj) and reduce_scatter (after o_proj) are introduced to enable sequence-parallel output aggregation while maintaining correctness under SFA CP.

### benchmark:
- TTFT increased by **527%**
- TPOT increased by **180%**

image

### Does this PR introduce _any_ user-facing change?
None
### How was this patch tested?

- vLLM version: v0.13.0
- vLLM main: https://github.com/vllm-project/vllm/commit/2f4e6548efec402b913ffddc8726230d9311948d

### CODE REVIEW

File: vllm_ascend/attention/sfa_v1.py [hunk 1]
```diff
@@ -221,12 +222,18 @@ def build(
...
                                                  (0, pad_size_slot),
                                                  value=-1)
+                slot_mapping_cp = torch.full(size=(num_tokens_per_device, ),
+                                             fill_value=-1,
+                                             dtype=slot_mapping.dtype,
+                                             device=slot_mapping.device)
+
             else:
                 slot_mapping = slot_mapping[:num_tokens_pad]
+                slot_mapping_cp = slot_mapping[local_start:local_end_with_pad]
```
[the remark this answers is not in the document]
Comment (zzhx1) [reply vllm_ascend/attention/sfa_v1.py:241]: @clrs97 please take a look at this issue

[the remark this answers is not in the document]
Comment (clrs97) [reply vllm_ascend/attention/sfa_v1.py:241]: `slot_mapping_cp` can either be `-1` or a slice of `slot_mapping` because it will be saved twice (first in `kv_rmsnorm_rope_cache`, second in `scatter_update`). Maybe setting to `-1` can be faster? But in graph mode slice is the only possible approach.

File: vllm_ascend/attention/sfa_v1.py [hunk 2]
```diff
@@ -906,7 +960,10 @@ def indexer_select_pre_process(
             k_pe, k_nope = torch.split(
                 k,
                 [self.qk_rope_head_dim, self.head_dim - self.qk_rope_head_dim],
-                dim=-1)  # [b,s,64+64]
+                dim=-1)
+
+            cos = cos.view(-1, 1, 1, self.qk_rope_head_dim)
```
Comment (clrs97) [inline vllm_ascend/attention/sfa_v1.py:963]: These same two lines in another function `indexer_select_post_process` can be removed now since those are not used.

**Comment by clrs97** [comment]
In linear_op.py:
```
def _get_row_parallel_op(
    prefix, layer
) -> Optional[Union[MLPRowParallelOp, OProjRowParallelOp,
                    Flashcomm2OProjRowParallelOp, MatmulAllreduceRowParallelOp,
                    SequenceRowParallelOp, ShardedCPRowParallelOp]]:
    if enable_dsa_cp() and "o_proj" in prefix:
        return ShardedCPRowParallelOp(layer)
```

Why is `enable_dsa_cp()` here instead of `enable_dsa_cp_with_layer_shard()`?

**Review by jianzs — COMMENTED** [review]
It's better to add test cases to make sure the behavior meets expectations—at least, include e2d tests.

File: vllm_ascend/attention/sfa_v1.py [hunk 3]
```diff
@@ -228,15 +230,18 @@ def build(
                 slot_mapping = nn.functional.pad(slot_mapping,
                                                  (0, pad_size_slot),
                                                  value=-1)
+                slot_mapping_cp = torch.full(size=(num_tokens_per_device, ),
+                                             fill_value=-1,
+                                             dtype=slot_mapping.dtype,
+                                             device=slot_mapping.device)
+
```
Comment (yiz-liu) [inline vllm_ascend/attention/sfa_v1.py]: Please make it a persistent buffer, otherwise accuracy issues may be introduced.

File: vllm_ascend/attention/sfa_v1.py [hunk 4]
```diff
@@ -374,33 +385,35 @@ def __init__(
...
+        self.n_head: int = self.indexer.n_head  # 64
+        self.head_dim: int = self.indexer.head_dim  # 128
+        self.wq_b = self.indexer.wq_b
+        self.wk = self.indexer.wk
+        self.weights_proj = self.indexer.weights_proj
+        self.k_norm = self.indexer.k_norm
+        self.cp_size = 1
+
+        self.enable_sfa_cp = enable_dsa_cp()
+        self.enable_sfa_cp_prefill_only = enable_dsa_cp_with_layer_shard()
```
Comment (yiz-liu) [inline vllm_ascend/attention/sfa_v1.py]: Another question I don't understand is that we are using `DSA` and `SFA` at the same time? Is there any nuances I don't notice? I think they are the same concept.

File: vllm_ascend/attention/sfa_v1.py [hunk 5]
```diff
@@ -746,6 +763,15 @@ def forward(
...
 
+        # all-gather o_proj weight for prefill stage of PD mix node
+        o_proj_full_handle = None
+        # if is PD mix stage, using original TP o_proj weight, and also need to full gather for o_proj weight for prefill stage.
+        should_shard_weight = self.enable_sfa_cp_prefill_only or attn_metadata.attn_state not in {
+            AscendAttentionState.DecodeOnly, AscendAttentionState.SpecDecoding
+        }
+
+        # TODO(zzzzwwjj): In sfa, prefill and decode have the same calculation formula,
+        # so `has_prefill` here is not necessary.
```
Comment (yiz-liu) [inline vllm_ascend/attention/sfa_v1.py]: Delete this.

File: vllm_ascend/attention/sfa_v1.py [hunk 6]
```diff
@@ -868,6 +909,21 @@ def forward(
...
+            # 2. decode: all-to-all the hidden_state before the o_proj forward.
+            result = self._handle_o_proj_weight_switch_and_forward(
+                attn_output=attn_output,
+                output=output,
+                output_padded=output_padded,
+                o_proj_full_handle=o_proj_full_handle,
+                should_shard_weight=should_shard_weight)
+            if result is output_padded:
+                return result
+            attn_output = result
```
Comment (yiz-liu) [inline vllm_ascend/attention/sfa_v1.py]: What's the difference between `result`, `output` and `output_padded`?

File: vllm_ascend/attention/sfa_v1.py (hunk 5 above)
Comment (rjg-lyh) [reply vllm_ascend/attention/sfa_v1.py]: Implemented in the latest code and the TODO has been removed. Please rebase and review again.

File: vllm_ascend/attention/sfa_v1.py (hunk 3 above)
Comment (clrs97) [reply vllm_ascend/attention/sfa_v1.py]: It is only in the intended prefill branch to potentially speed up kv saving, in decode graph branch it won't create new tensor. However, it turned out the kv saving can not be speeded up by this, so I suggest remove this and fallback to `slot_mapping_cp = slot_mapping[local_start:local_end_with_pad]`

**Comment by starmountain1997** [comment]
> It's better to add test cases to make sure the behavior meets expectations—at least, include e2d tests.

Once PR #5702 is merged, I will merge CI PR #5921. The test cases in it cover this scenario.

File: vllm_ascend/attention/sfa_v1.py (hunk 6 above)
Comment (zzhx1) [reply vllm_ascend/attention/sfa_v1.py]: I realized that this expression was not accurate, and I refactored it here.
Use `require_o_proj_forward` to determine whether to return directly

File: vllm_ascend/attention/sfa_v1.py [hunk 7]
```diff
@@ -873,6 +908,20 @@ def forward(
                            dependency=attn_output,
                            max_size=MAX_O_PROJ_PREFETCH_SIZE,
                            enabled=self.enable_prefetch)
+
+        if self.enable_dsa_cp and not self.enable_dsa_cp_prefill_only:
+            # When using SFA-CP with pd mixed, o_proj has two cases:
+            # 1. prefill: o_proj is a TP weight, we need to all-gather o_proj weight to switch TP=1.
+            # 2. decode: all-to-all the hidden_state before the o_proj forward.
+            result, require_o_proj_forward = self._handle_o_proj_weight_switch_and_forward(
```
Comment (jianzs) [inline vllm_ascend/attention/sfa_v1.py:916]: Where is `_handle_o_proj_weight_switch_and_forward` defined?

File: vllm_ascend/attention/sfa_v1.py [hunk 8]
```diff
@@ -981,3 +1030,92 @@ def indexer_select_post_process(
...
+        For PD mix stage:
+        - Use original TP o_proj weight for decode phase
+        - Need full-gather o_proj weight from all TP ranks for prefill phase
+        """
+        if AscendSFAImpl.o_proj_full_pool is None:
+            sample = self.o_proj.weight
+            AscendSFAImpl.o_proj_full_pool = torch.empty(
+                (sample.shape[0] * self.tp_size, sample.shape[1]),
+                dtype=sample.dtype,
+                device=sample.device)
```
Comment (jianzs) [inline vllm_ascend/attention/sfa_v1.py:1048]: If we don't enable layer sharding, do we still need to initialize the tensor?

File: vllm_ascend/attention/sfa_v1.py [hunk 9]
```diff
@@ -981,3 +1030,92 @@ def indexer_select_post_process(
...
+
+        # Precompute Full-mode quantization parameters by repeating TP parameters across all TP ranks
+        self.o_proj_full_aclnn_input_scale = self.o_proj.aclnn_input_scale.repeat(
+            self.tp_size)
+        self.o_proj_full_aclnn_input_scale_reciprocal = self.o_proj.aclnn_input_scale_reciprocal.repeat(
+            self.tp_size)
+        self.o_proj_full_aclnn_input_offset = self.o_proj.aclnn_input_offset.repeat(
+            self.tp_size)
+
+    def x_handle_o_proj_weight_switch_and_forward(
```
Comment (jianzs) [inline vllm_ascend/attention/sfa_v1.py]: remove the prefix `x` ?

File: vllm_ascend/attention/sfa_v1.py [hunk 10]
```diff
@@ -981,3 +1030,92 @@ def indexer_select_post_process(
...
+        Handle o_proj weight switching between TP-mode and Full-mode, and execute forward computation.
+        """
+        # Gather o_proj weight from all TP ranks for Full-mode computation
+        if should_shard_weight:
+            # Wait for the completion of o_proj weight all-gather operation
+            if o_proj_full_handle is not None:
+                o_proj_full_handle.wait()
+
+            # Switch o_proj to Full-mode (gathered weight from all TP ranks)
+            self.o_proj.weight.set_(AscendSFAImpl.o_proj_full_pool)
```
Comment (jianzs) [inline vllm_ascend/attention/sfa_v1.py:1088]: To avoid these two copy operations, it's better to use separate variables for full-mode and tp-mode.

File: vllm_ascend/attention/sfa_v1.py (hunk 8 above)
Comment (zzhx1) [reply vllm_ascend/attention/sfa_v1.py:1048]: This statement has already been judged in the upper function; initialization will only occur if layer sharding is enabled.
```bash
if self.enable_dsa_cp:
            if self.enable_dsa_cp_prefill_only:
                for layer in (self.layer_sharding_kwargs or []):
                    if is_hidden_layer(layer):
                        post_process_after_loading_for_shard_weight_series(
                            layer)
            else:
                self._init_o_proj_tp_full_params()
```

**Review by jianzs — APPROVED** [review]

**Review by yiz-liu — APPROVED** [review]
